content/blog/en/introduction-to-testing-with-dojo.md

```diff
@@ -0,0 +1,151 @@
+
+Of course there are a number of different types of testing that help to ensure that an application has fewer bugs but for the purpose of this blog we'll be concentrating on unit testing.
+
+When creating a Dojo application using the Dojo [cli-create-app](https://dojo.io/blog/dojo-cli-template-app) command, as part of the project there are a set of unit tests that we hope will help guide everyone in unit testing your application.
+
+![test structure](assets/blog/introduction-to-testing-with-dojo/test-structure.png)
+
+## Running Tests
+
+The tests are run using the Dojo cli with the `@dojo/cli-test-intern` command that is preinstalled with the scaffolded application.
+
+```bash
+dojo test
+```
+
+You can also run the tests using the predefined scripts in the `package.json`.
+
+```bash
+npm test
+```
+
+The results should be output to the terminal
+
+![test results](assets/blog/introduction-to-testing-with-dojo/test-results.png)
+
+So what does one of these tests look like?
+
+> tests/unit/widgets/Profile.tsx
+
+```tsx
+const { describe, it } = intern.getInterface('bdd');
+import harness from '@dojo/framework/testing/harness';
+import { tsx } from '@dojo/framework/core/vdom';
+
+import Profile from '../../../src/widgets/Profile';
+import * as css from '../../../src/widgets/styles/Profile.m.css';
+
+describe('Profile', () => {
+	it('default renders correctly', () => {
+		const h = harness(() => <Profile username="Dojo User" />);
+		h.expect(() => <h1 classes={[css.root]}>Welcome Dojo User!</h1>);
+	});
+});
+```
+
+Dojo comes with a [testing harness](https://dojo.io/learn/testing/dojo-test-harness) for testing your application's widgets. The harness enables you to shallowly assert a widget's render output matches the expected output based on the widget's properties. Ideally, rendering a widget is going to be a _[pure function](https://github.com/MostlyAdequate/mostly-adequate-guide/blob/master/ch03.md#chapter-03-pure-happiness-with-pure-functions)_, meaning they should be pretty easy to test when given the same input. 
+
+## Assertion Templates
+
+The `Profile` widget we tested above has a property username that we can test against in the output. We could rewrite the entirety of the expected render output for each test (that's a lot of typing) or we could create an assertion to test against that would allow us to change the expected properties on each run. An important aspect of assertion templates is they ensure assertions are made for entire unit, which in our case is always the entire render output. It can often temping be to trigger an action and then only partially assert the section of the render output that the action is expected to affect. Partially asserting a widget's render output can be problematic as it means a test can "pass" but in fact the render output outside of the section asserted was affected and not caught.
```

Comment:
```suggestion
The `Profile` widget tested above has the `username` property that we can test against in the output. We could rewrite the entirety of the expected render output for each test (that's a lot of typing) or we could create an assertion to test against that would allow us to change the expected properties on each run. An important aspect of assertion templates is they ensure assertions are made for an entire unit, which in our case is always the entire render output. It can be tempting be to trigger an action and then only partially assert the section of the render output that the action is expected to affect. Partially asserting a widget's render output can be problematic as it means a test can "pass", but in fact the render output outside of the section asserted was affected and not caught, leading to uncaught bugs in a widget's structure. This means that we would not know if the action of triggering the event handler had any unexpected affects on other areas of a widget's render output.
```